pkg/agent/kubernetes.go

```diff
@@ -94,17 +97,166 @@ func patchNodeStatus(cli *k8sclient.Clientset, fields map[string]string) error {
-	listOpts := meta_v1.ListOptions{
-		FieldSelector: "metadata.name=" + name,
+// watch is a wrapper around the k8s watch.Interface
+type watch struct {
+	parent  *watcher
+	kind    string
+	ns      namespace
+	name    string
+	openfn  func(namespace, string) (k8swatch.Interface, error)
+	queryfn func(namespace, string) (interface{}, error)
+	stop    chan struct{}
+	events  chan k8swatch.Event
+}
+
+type openFn func(ns namespace, name string) (k8swatch.Interface, error)
+type queryFn func(ns namespace, name string) (interface{}, error)
+
+const (
+	// SyntheticMissing is a synthetic initial event for currently non-existent object.
+	SyntheticMissing = k8swatch.EventType("SyntheticMissing")
+)
+
+func newWatch(parent *watcher, kind string, ns namespace, open openFn, query queryFn) *watch {
+	return &watch{
+		parent:  parent,
+		kind:    kind,
+		ns:      ns,
+		stop:    make(chan struct{}),
+		events:  make(chan k8swatch.Event),
+		openfn:  open,
+		queryfn: query,
 	}
-	watch, err := cli.CoreV1().ConfigMaps(ns).Watch(listOpts)
-	if err != nil {
-		return nil, agentError("failed to create a watch for configmaps in ns %q: %v", ns, err)
+}
+
+// configNodeWatch creates a watch for k8s Node
+func newNodeWatch(parent *watcher) *watch {
+	w := newWatch(parent, "Node", namespace(""),
+		func(ns namespace, name string) (k8swatch.Interface, error) {
+			selector := meta_v1.ListOptions{FieldSelector: "metadata.name=" + name}
+			k8w, err := parent.k8sCli.CoreV1().Nodes().Watch(selector)
+			if err != nil {
+				return nil, err
+			}
+			return k8w, nil
+		},
+		func(ns namespace, name string) (interface{}, error) {
+			noopts := meta_v1.GetOptions{}
+			node, err := parent.k8sCli.CoreV1().Nodes().Get(name, noopts)
+			if err != nil {
+				return nil, err
+			}
+			return node, nil
+		})
+	w.Start(nodeName)
+	return w
+}
+
+// configMapWatch creates a watch for k8s ConfigMap
+func newConfigMapWatch(parent *watcher, name string, ns namespace) *watch {
+	w := newWatch(parent, "ConfigMap", ns,
+		func(ns namespace, name string) (k8swatch.Interface, error) {
+			selector := meta_v1.ListOptions{FieldSelector: "metadata.name=" + name}
+			k8w, err := parent.k8sCli.CoreV1().ConfigMaps(string(ns)).Watch(selector)
+			if err != nil {
+				return nil, err
+			}
+			return k8w, nil
+		},
+		func(ns namespace, name string) (interface{}, error) {
+			noopts := meta_v1.GetOptions{}
+			cm, err := parent.k8sCli.CoreV1().ConfigMaps(string(ns)).Get(name, noopts)
+			if err != nil {
+				return nil, err
+			}
+			return cm, nil
+		})
+	w.Start(name)
+	return w
+}
+
+func (w *watch) Name() string {
+	ns, name := w.ns, w.name
+	if ns != "" {
+		ns += "/"
+	}
+	if name == "" {
+		name = "<none>"
 	}
+	return w.kind + ":" + string(ns) + name
+}
+
+// Query queries the object being watched.
+func (w *watch) Query() (interface{}, error) {
+	if w.name == "" {
+		return nil, nil
+	}
+	return w.queryfn(w.ns, w.name)
+}
+
+// Start restart watching a potentially new object.
```

Comment:
"Start restart"?

Reply:
Fixed comment. It used both to start and restart a watch.